sw/host/tests/crypto/aes_gcm_nist_kat/src/main.rs

```diff
@@ -0,0 +1,231 @@
+        aad_length: test_case.aad.len(),
+        tag: ArrayVec::try_from(test_case.tag.as_slice()).unwrap(),
+        tag_length: test_case.tag.len(),
+        input: ArrayVec::try_from(input_data.as_slice()).unwrap(),
+        input_length: input_data.len(),
+    }
+    .send(&*uart)?;
+
+    // Get the output
+    let gcm_output = CryptotestAesGcmOutput::recv(&*uart, opts.timeout, false)?;
+
+    let mut failed = false;
+    vec![("oneshot", gcm_output.oneshot_output),
+         ("stepwise", gcm_output.stepwise_output)]
+    .into_iter()
+    .for_each(|(mode, output)|{
+        let target_output = match test_case.operation.as_str() {
+            "encrypt" => &test_case.ciphertext,
+            "decrypt" => &test_case.plaintext,
+            _ => panic!("Invalid operation type")
+        };
+        let success = if target_output.len() > gcm_output.output_length {
+            // If we got a shorter digest back then the test asks for, we
+            // can't accept the digest, even if the beginning bytes match.
+            false
+        } else {
+            // Some test cases only specify the beginning bytes of the
+            // expected digest, so we only check up to what the test
+            // specifies.
+            target_output[..] == output[..target_output.len()]
+        };
+        if target_output.len() > 0 && test_case.result != success {
+            log::info!(
+                "FAILED {} test #{} in {} mode: expected = {}, actual = {}",
+                test_case.operation,
+                test_case.test_case_id,
+                mode,
+                test_case.result,
+                success
+            );
+            failed = true;
+        }
+    });
+    if failed {
+        *fail_counter += 1;
+    }
+    Ok(())
+}
+
+fn is_input_empty(aes_gcm_test: &AesGcmTestCase) -> bool {
+    if aes_gcm_test.operation.as_str() == "encrypt" && aes_gcm_test.plaintext.len() == 0 {
+        return true;
+    }
+    if aes_gcm_test.operation.as_str() == "decrypt" && aes_gcm_test.ciphertext.len() == 0 {
+        return true;
+    }
+    false
+}
+
+fn test_aes_gcm(opts: &Opts, transport: &TransportWrapper) -> Result<()> {
+    let uart = transport.uart("console")?;
+    uart.set_flow_control(true)?;
+    let _ = UartConsole::wait_for(&*uart, r"Running [^\r\n]*", opts.timeout)?;
+
+    let mut test_counter = 0u32;
+    let mut fail_counter = 0u32;
+    let test_vector_files = &opts.aes_gcm_json;
+    for file in test_vector_files {
+        let raw_json = fs::read_to_string(file)?;
+        let aes_gcm_tests: Vec<AesGcmTestCase> = serde_json::from_str(&raw_json)?;
+        for aes_gcm_test in &aes_gcm_tests {
+            test_counter += 1;
+            log::info!("Test counter: {}", test_counter);
+            if is_input_empty(&aes_gcm_test) {
+                continue;
+            }
+            run_aes_gcm_testcase(aes_gcm_test, opts, transport, &mut fail_counter)?;
+        }
+    }
+    assert_eq!(
```

Comment:
Just FYI: I change the output format for the test harnesses in #21923, that way it's possible to see which tests failed in CI while still suppressing all the UART output in Azure pipelines. Nothing there should block this PR, but if you're so inclined you can update this test harness to use the format we use over there to avoid having to rebase later :)

Reply:
I think I will go with the new test harness format. Thanks for the heads up!